client/model/ApplicationStateModel.js

```diff
@@ -123,35 +126,135 @@ define(
+                opdep_ajax.success(function (data) {
+                    opdepArray = data.opdep //gets the array from dict
+                    //double for loop
+                    ko.utils.arrayForEach(self.applicationStateArray(), function (item) {
+                        opdepArray.map(function (componentId) {
+                            if (item.componentId === componentId.replace('/spot/software/state/application/', '')) {
+                                //Add to array if the element hasn't been added
+                                if ($.inArray(item, self.opdepAppStateArray()) == -1 ){
+                                    self.opdepAppStateArray().push(item)
+                                }
+                            }
+                        })
+                    })
+                    if (execute_command){
+                        swal({
+                            title: 'ARE YOU SURE?!',
+                            text: self.path_message_appstate(),
+                            type: "warning",
+                            showCancelButton: true,
+                            confirmButtonColor: "#DD6B55",
+                            confirmButtonText: "Yes, Restart them!",
+                            closeOnConfirm: true
+                        },
+                        function(isConfirm){
+                            if (isConfirm) {
+                                // check if previous async call is completed before any of this runs
+                                self.executeOpdepControl({'com': 'ignore', 'clear_group': true});
+                                self.executeOpdepControl({'com': 'stop', 'stay_down': false, 'clear_group': true});
+                                self.checkStopped();
+                            } else {
+                                return;
+                            }
+                        })
+                    }
+                })
+            };
+
+            self.createOpdepStateArray = function(){
+                var opdep_ajax;
+                if (self.groupMode()){
+                    //iterate groupcontrol and create array
+                    for (i = 0; i < self.groupControl().length; i++){
+                        ApplicationState = self.groupControl()[i]
+                        opdep_ajax = self.OpdepAjax(ApplicationState.componentId)
+                        if (i == (self.groupControl().length - 1)){
+                            self.addtoOpDepArray(opdep_ajax, true)
+                        }
+                        else{
+                            self.addtoOpDepArray(opdep_ajax, false)
+                        }
+
+                    }
+                }
+                else{
+                    opdep_ajax = self.OpdepAjax(self.clickedApp().componentId)
+                    self.addtoOpDepArray(opdep_ajax, true)
+                }
+            };
+
+            self.executeOpdepControl = function(options) {
```

Comment:
You _might_ be able to reuse executeGroupControl by changing it to accept an observable array instead of duplicating it here...?